src/controllers/hid/hidiooutputreport.cpp

```diff
@@ -0,0 +1,155 @@
+        // Only HID Feature items may be attributes of other items.
+        // This means there is always a one to one relationship to the state of control(s)/LED(s),
+        // and if the state is not changed, there's no need to execute the time consuming hid_write again.
+
+        // Setting m_possiblyUnsentDataCached to false prevents,
+        // that the byte array compare operation is executed for the same data again
+        m_possiblyUnsentDataCached = false;
+
+        reportCacheLock.unlock();
+
+        qCDebug(logOutput) << "t:" << startOfHidWrite.formatMillisWithUnit()
+                           << " Skipped identical Output Report for"
+                           << deviceInfo.formatName() << "serial #"
+                           << deviceInfo.serialNumber() << "(Report ID"
+                           << m_reportId << ")";
+
+        // Return with false, to signal the caller, that no time consuming IO operation was necessary
+        return false;
+    }
+
+    // Preemtive set m_lastSentData and m_possiblyUnsentDataCached,
+    // to release the mutex during the time consuming hid_write operation.
+    // In the unlikely case that hid_write fails, they will be invalidated afterwards
+    // This is safe, because these members are only reset in this scope of this method,
+    // and concurrent execution of this method is prevented by locking pHidDeviceMutex
+    m_lastSentData.swap(m_cachedData);
+    m_possiblyUnsentDataCached = false;
+
+    reportCacheLock.unlock();
+
+    auto hidDeviceLock = lockMutex(pHidDeviceMutex);
+
+    // hid_write can take several milliseconds, because hidapi synchronizes
+    // the asyncron HID communication from the OS
+    int result = hid_write(pHidDevice,
+            reinterpret_cast<const unsigned char*>(m_lastSentData.constData()),
+            m_lastSentData.size());
+    if (result == -1) {
+        qCWarning(logOutput) << "Unable to send data to" << deviceInfo.formatName() << ":"
+                             << mixxx::convertWCStringToQString(
+                                        hid_error(pHidDevice),
+                                        kMaxHidErrorMessageSize);
+    }
+
+    hidDeviceLock.unlock();
+
+    if (result == -1) {
+        reportCacheLock.relock();
+        // Clear the m_lastSentData because the last send data are not reliable known.
+        // These error should not occur in normal operation,
+        // therefore the performance impact of additional memory allocation
+        // at the next call of this method is negligible
+        m_lastSentData.clear();
+        m_lastSentData.append(m_reportId);
+        m_possiblyUnsentDataCached = true;
+
+        // Return with true, to signal the caller, that the time consuming hid_write operation was executed
+        // (Note, that the return value isn't an error code)
+        return true;
+    }
+
+    qCDebug(logOutput) << "t:" << startOfHidWrite.formatMillisWithUnit() << " "
```

Comment:
m_cachedDataMutex is still logged while generating this message. Please unlock it before.

Reply:
I don't think so, it's unlocked in line 114. And the relock is in the if branch, which returns early.